include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,1025 @@
+   */
+  template <int dim, int n_components, typename value_type_>
+  struct PrecomputedEvaluationData
+  {
+    using value_type = typename internal::FEPointEvaluation::
+      EvaluatorTypeTraits<dim, n_components, value_type_>::value_type;
+
+    using gradient_type = typename internal::FEPointEvaluation::
+      EvaluatorTypeTraits<dim, n_components, value_type_>::gradient_type;
+
+    /**
+     * Values at quadrature points.
+     */
+    AlignedVector<value_type> values;
+
+    /**
+     * Gradients at quadrature points.
+     */
+    AlignedVector<gradient_type> gradients;
+  };
+
+  /**
+   * A class that stores a CRS like structure to access
+   * PrecomputedEvaluationData. In most cases a one-level CRS structure
+   * is enough. In this case @c ptrs has to be constructed and the shift can be
+   * obtained with `get_shift(index)`. The field @c ptrs_ptrs stays empty.
+   * It is only filled if a two level structure is needed. In this case
+   * `get_shift(cell_index, face_number)` return the correct shift.
+   */
+  struct PrecomputedEvaluationDataView
+  {
+    /**
+     * Get a pointer to data at @p index.
+     */
+    unsigned int
+    get_shift(const unsigned int index) const;
+
+    /**
+     * Get a pointer to data at (@p cell_index, @p face_number).
+     */
+    unsigned int
+    get_shift(const unsigned int cell_index,
+              const unsigned int face_number) const;
+
+    /**
+     * Get the number of stored values.
+     */
+    unsigned int
+    size() const;
+
+    /**
+     * This parameter can be used if indices do not start with 0.
+     */
+    unsigned int start = 0;
+
+    /**
+     * Pointers to @c ptrs.
+     */
+    std::vector<unsigned int> ptrs_ptrs;
+
+    /**
+     * Pointers to data.
+     */
+    std::vector<unsigned int> ptrs;
+  };
+
+  /**
+   * A class helps to access PecomputedEvaluationData in a thread-safe
+   * manner. Each thread has to create this wrapper class on its own to
+   * avoid race-conditions during reinit().
+   */
+  template <int dim, int n_components, typename value_type_>
+  class PrecomputedEvaluationDataAccessor
+  {
+    using value_type =
+      typename PrecomputedEvaluationData<dim, n_components, value_type_>::
+        value_type;
+    using gradient_type =
+      typename PrecomputedEvaluationData<dim, n_components, value_type_>::
+        gradient_type;
+
+  public:
+    /**
+     * Constructor.
+     */
+    PrecomputedEvaluationDataAccessor(
+      const PrecomputedEvaluationData<dim, n_components, value_type_> &data,
+      const PrecomputedEvaluationDataView                             &view);
+
+    /**
+     * Get the value at quadrature point @p q. The entity on which the values
+     * are defined is set via `reinit()`.
+     */
+    const value_type
+    get_value(const unsigned int q) const;
+
+    /**
+     * Get the gradients at quadrature point @p q. The entity on which the
+     * gradients are defined is set via `reinit()`.
+     */
```

Comment:
I always forget our convention with respect to doxygen, but do you want to add the "@param" field?